### ISSUE TITLE
Control of implicit assignment index variable

### ISSUE THREAD

**Issue by cnuernber** [dmlc/tvm#972]
In halide, functions expressed are (mainly) pure.  Implementationwise, however, there is an implicit assignment happening under the covers.

tvm extends halide with the notion of a n-dimensional tensor where each dimension has a byte-wise stride associated with it (either explicitly or implicitly).  

In cortex I believe I came up with a generalization of the python broadcast semantics along with a simpler way of describing the elementwise operations (assignment, unary, binary, and ternary opterations).  I would like to try to support this generalization in TVM but it requires control over the assignment index variable.

This is what I mean:

[assign-index value] = operation(x,y,z,w)
dest[assign-index] = value.

This allows me to do things like broadcasting where the shape of the destination is taken into account (specifically broadcasting a vector into the rows or columns of a matrix with an assignment call).  It also means that all of the elementwise operations are described using an num-elems argument which is the logical number of elements that the operation's broadcast shape has.  It also allows me to perform and describe operations that are tensor operations but keep the underlying backing store 1-dimensional vectors (which I know is not a big deal but it is more flexibility).

So binary operation looks like:

a = b op c.

Regardless of the dimensionality of a,b,c, it was transmitted to cuda like (binary-op! a b c op num-elems) and I let one function that could convert from elem-index->tensor-buffer-index via the tensor's shape and the broadcast shape of the operation do all the heavy lifting on all sides of the equation.  This is slow if that function isn't specialized for simple cases but with TVM I can use dynamic specializations of the function and produce new binary-op functions based on specific combinations of shape and the max-shape of the operation.  This is my hope, anyway.  

Basically in most of the tvm examples I have seen the destination is implicit but TVM itself has figured out how to perform an operation that arrives at an index value by going through the tensor's shape and strides (and byte_offset) after halide does it's work.  

So, precisely I want to say something like:

>>perform the (non-reducing-or-scanning) operation over Y dimensions but at the end here is a transformation from the Y indices to one final index to use when writing the data.

Can I override or change this function as a user or do I need to limit my operations to always broadcasting specifically to the the shape of the result thus allowing (most of) the tvm system to work unchanged?

**Comment by tqchen** [comment]
If i understand correctly, you are asking for a specific scatter operation with given index operations. For arbitrary flexible about the final index specification, we might need to come up with a new primitive other than compute. TVM is designed to be flexible on that, and we could possibly come up with a operator def like those in scan.

Supporting arbitrary index scan order, however, can have impact on bound inference and workload partitioning on GPU. That is why usually the "pull" style compute is preferred.

On the other hand, the thing that you ask for, like storing things back to a one dimensional vector store implicitly, can already be supported. The idea is that the Buffer is a symbolic definition of the Tensor. It does not have to correspond to a real Tensor object that has strides, but instead it only declares how the variables maps to the Tensor object. Since this is done at compile time, different Buffer can give us different specialization of NDArray data structure, as long as we specifies how we could map it to generic Buffer.

Finally, it might be helpful to take look at topi, which already contains support for numpy style broadcasting

**Comment by cnuernber** [comment]
Yes, possibly scatter actually thinking about it possibly reduction which of course can't work.  

Thanks for the pointer to the buffer object; that is very helpful.

I am researching topi and no need to really consider anything new; I am just trying to figure out how all the pieces map to concrete operations and what actual bounds there are.

This is plenty for now for me to chew on.
